### PR TITLE
Lpal 185 aurora dev upgrade

### PR DESCRIPTION
**Pull request by williamfalconeruk**
## Purpose
To implement dev environment with aurora Serverless.
This will open up the migration to Aurora in preproduction and production as the code will be in place in readiness for that.

Fixes LPAL-185

## Approach

- Take example module from digideps /sirius infra and adjust for needs on make.
- Configure so it is only available in dev.
- Account level changes to support specific db subnet group.

Follow on work will be to tier out db into it's own subnets, rather than using private based subnet.

## Learning

- Terraform modules
- VPC, DB Subnet groups and subnets

## Checklist

* [x] I have performed a self-review of my own code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [ ] I have added tests to prove my work
* [ ] The product team have tested these changes

### CODE REVIEW

**Review by charlesmmarshall — DISMISSED** [review]

File: terraform/environment/terraform.tfvars.json [hunk 1]
```diff
@@ -6,14 +6,19 @@
   },
   "accounts": {
     "development": {
+      "deletion_protection" :false,
+      "aurora_enabled" : true,
+      "aurora_serverless" : true,
+      "aurora_instance_count" : 1,
+      "always_on": false,
```
Comment (TomTucka) [inline terraform/environment/terraform.tfvars.json:13]: You don't both `aurora_enabled` & `always_on `. They do the same thing

File: terraform/environment/rds.tf [hunk 2]
```diff
@@ -73,3 +103,14 @@ resource "aws_security_group_rule" "rds-api" {
   source_security_group_id = aws_security_group.rds-client.id
   security_group_id        = aws_security_group.rds-api.id
 }
+
+locals {
```
[the remark this answers is not in the document]
Comment (TomTucka) [reply terraform/environment/rds.tf:107]: I would do this in the task definition.

**Review by TomTucka — DISMISSED** [review]

File: terraform/account/network.tf [hunk 3]
```diff
@@ -94,3 +94,18 @@ resource "aws_route" "private" {
   destination_cidr_block = "0.0.0.0/0"
   nat_gateway_id         = element(aws_nat_gateway.nat.*.id, count.index)
 }
+
+
+resource "aws_db_subnet_group" "data_persistence" {
+  name       = "data-persistence-subnet-${local.account.stack_name}"
```
Comment (andrewpearce-digital) [inline terraform/account/network.tf:100]: This name-spacing isn't required here as there's only one "stack" (vpc) per account, and only one db_subnet_group.

File: terraform/account/terraform.tfvars.json [hunk 4]
```diff
@@ -10,21 +10,25 @@
       "is_production": "false",
       "front_certificate_domain_name": "*.front.lpa.opg.service.justice.gov.uk",
       "admin_certificate_domain_name": "*.admin.lpa.opg.service.justice.gov.uk",
-      "retention_in_days": 7
+      "retention_in_days": 7,
+      "stack_name" : "dev-vpc"
```
Comment (andrewpearce-digital) [inline terraform/account/terraform.tfvars.json:14]: Given how it's used, I don't think this is necessary. If a namespacing variable was needed account_name.

**Review by andrewpearce-digital — DISMISSED** [review]

File: terraform/environment/rds.tf [hunk 5]
```diff
@@ -17,11 +24,34 @@ resource "aws_db_instance" "api" {
...
   backup_retention_period     = local.account.backup_retention_period
-  deletion_protection         = local.account.prevent_db_destroy
+  deletion_protection         = local.account.deletion_protection
   tags                        = local.default_tags
   allow_major_version_upgrade = true
 }
 
+module "api_aurora" {
+  source                        = "./modules/aurora"
+  count                         = local.account_name == "development" ? 1 : 0
```
[the remark this answers is not in the document]
Comment (TomTucka) [reply terraform/environment/rds.tf:34]: `local.account.aurora_enabled ? 1 : 0` would be much better here I think.

File: terraform/account/terraform.tfvars.json (hunk 4 above)
Comment (TomTucka) [reply terraform/account/terraform.tfvars.json:14]: `stack_name` is definitely a hangover from Sirius. You can engineer this out with little effort

File: terraform/environment/rds.tf [hunk 6]
```diff
@@ -17,11 +24,34 @@ resource "aws_db_instance" "api" {
...
+  aurora_serverless             = local.account.aurora_serverless
+  account_id                    = data.aws_caller_identity.current.account_id
+  apply_immediately             = ! local.account.deletion_protection
+  cluster_identifier            = "api2"
+  db_subnet_group_name          = "data-persistence-subnet-${local.account.vpc_name}"
```
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/environment/rds.tf:39]: proposing to change this to `data-persistence-subnet-default`?

File: terraform/account/network.tf (hunk 3 above)
Comment (williamfalconeruk) [reply terraform/account/network.tf:100]: will change based on above proposal.

File: terraform/account/terraform.tfvars.json (hunk 4 above)
Comment (williamfalconeruk) [reply terraform/account/terraform.tfvars.json:14]: will remove, and adjust.

File: terraform/environment/rds.tf (hunk 5 above)
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/environment/rds.tf:34]: this was intentional, as a secondary protection, will unhard code when we're ready to proceed to pre and production.

[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/environment/rds.tf:34]: will change.

File: terraform/environment/rds.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/environment/rds.tf:107]: we've 2 task definitions that rely on this. This is a stopgap until we've migrated pre and production to aurora instances. that work is to follow. at that point happy to follow the pattern that sirius is using for this, as the module will be managing the instances.

[the remark this answers is not in the document]
Comment (williamfalconeruk) [reply terraform/environment/rds.tf:107]: moving this to the locals.tf for now.

**Review by charlesmmarshall — APPROVED** [review]

**Pull request merged**